### PR TITLE
llvm 15, 16: add zstd support

### PR DESCRIPTION
**Pull request by SaltyKitkat**
###### Description of changes

Since llvm 16 support using zstd as its compression method, we should also support that.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by rrbutani** [comment]
(for anyone following along, this is another attempt at #238266)

File: pkgs/development/compilers/llvm/16/llvm/default.nix [hunk 1]
```diff
@@ -383,7 +385,10 @@ in
   checkTarget = "check-all";
 
   # For the update script:
-  passthru.monorepoSrc = monorepoSrc;
+  # passthru.monorepoSrc = monorepoSrc;
```
Comment (rrbutani) [inline pkgs/development/compilers/llvm/16/llvm/default.nix:388]:
```suggestion
```

**Review by rrbutani — CHANGES_REQUESTED** [review]
Thanks for continuing to work on this; a few more things:
  - [x] can you change your commit to start with `llvm_16` instead of `llvm 16`? ofborg picks up the former and it's convention
  - [x] as [mentioned in #238266 LLVM 16+](https://github.com/NixOS/nixpkgs/pull/238266#pullrequestreview-1484846591) has a [`LLVM_USE_STATIC_ZSTD` option](https://github.com/llvm/llvm-project/commit/fc1da043f4f9198303abd6f643cf23439115ce73); given the direction #237563 is headed (i.e. potentially having static/dynamic library outputs present with `isStatic = false`/`isStatic = true` respectively IIUC) I think it'd be good to add something like this to `cmakeFlags`:
    ```nix
    ++ lib.optional enableZstd "-DLLVM_USE_STATIC_ZSTD=${toString stdenv.hostPlatform.isStatic}"
    ```
  - [x] as mentioned, now that we're targeting `staging` anyways we can apply these changes (minus the `LLVM_USE_STATIC_ZSTD` bit) to `llvmPackages_15` in this PR

File: pkgs/development/compilers/llvm/16/llvm/default.nix [hunk 2]
```diff
@@ -383,7 +385,10 @@ in
...
+  passthru = {
+    inherit monorepoSrc enableZstd;
+  };
```
Comment (rrbutani) [inline pkgs/development/compilers/llvm/16/llvm/default.nix:391]:
```suggestion
  passthru = { inherit enableZstd; };
```
(see above)

Comment (SaltyKitkat) [reply pkgs/development/compilers/llvm/16/llvm/default.nix:390]: So it's unnecessary to set `passthru.monorepoSrc` here?

**Comment by SaltyKitkat** [comment]
> Thanks for continuing to work on this; a few more things:
> 
>     * [x]  can you change your commit to start with `llvm_16` instead of `llvm 16`? ofborg picks up the former and it's convention
> 
>     * [ ]  as [mentioned in #238266 LLVM 16+](https://github.com/NixOS/nixpkgs/pull/238266#pullrequestreview-1484846591) has a [`LLVM_USE_STATIC_ZSTD` option](https://github.com/llvm/llvm-project/commit/fc1da043f4f9198303abd6f643cf23439115ce73); given the direction [bzip2, zstd: Add `enableStatic` option #237563](https://github.com/NixOS/nixpkgs/pull/237563) is headed (i.e. potentially having static/dynamic library outputs present with `isStatic = false`/`isStatic = true` respectively IIUC) I think it'd be good to add something like this to `cmakeFlags`:
>       ```nix
>       ++ lib.optional enableZstd "-DLLVM_USE_STATIC_ZSTD=${toString stdenv.hostPlatform.isStatic}"
>       ```
> 
>     * [ ]  as mentioned, now that we're targeting `staging` anyways we can apply these changes (minus the `LLVM_USE_STATIC_ZSTD` bit) to `llvmPackages_15` in this PR

**Comment by SaltyKitkat** [comment]
Is it necessary to mention this change in release note or something like that?

**Comment by SaltyKitkat** [comment]
> Thanks for continuing to work on this; a few more things:
> 
> * [x]  can you change your commit to start with `llvm_16` instead of `llvm 16`? ofborg picks up the former and it's convention
> 
> * [x]  as [mentioned in #238266 LLVM 16+](https://github.com/NixOS/nixpkgs/pull/238266#pullrequestreview-1484846591) has a [`LLVM_USE_STATIC_ZSTD` option](https://github.com/llvm/llvm-project/commit/fc1da043f4f9198303abd6f643cf23439115ce73); given the direction [bzip2, zstd: Add `enableStatic` option #237563](https://github.com/NixOS/nixpkgs/pull/237563) is headed (i.e. potentially having static/dynamic library outputs present with `isStatic = false`/`isStatic = true` respectively IIUC) I think it'd be good to add something like this to `cmakeFlags`:
>       ```nix
>       ++ lib.optional enableZstd "-DLLVM_USE_STATIC_ZSTD=${toString stdenv.hostPlatform.isStatic}"
>       ```
> 
> * [x]  as mentioned, now that we're targeting `staging` anyways we can apply these changes (minus the `LLVM_USE_STATIC_ZSTD` bit) to `llvmPackages_15` in this PR

Besides, I set CmakeFlags `LLVM_PREFER_STATIC_ZSTD` in llvm 15 due to [this](https://github.com/llvm/llvm-project/commit/fc1da043f4f9198303abd6f643cf23439115ce73#diff-f6c3fac0de0e2bbd79ac93b0a6aabe77d9961c5487347a2170c78ac77770d8af)

File: nixos/modules/services/printing/cupsd_tmpfiles.conf
Comment (rrbutani) [inline nixos/modules/services/printing/cupsd_tmpfiles.conf:1]: I think this was added by mistake?

**Comment by rrbutani** [comment]
> Is it necessary to mention this change in release note or something like that?

I think that's not necessary; this isn't a breaking change and `enableZstd` is enabled by default (users won't have to do anything to get the feature).

---

> Besides, I set CmakeFlags `LLVM_PREFER_STATIC_ZSTD` in llvm 15 due to [this](https://github.com/llvm/llvm-project/commit/fc1da043f4f9198303abd6f643cf23439115ce73#diff-f6c3fac0de0e2bbd79ac93b0a6aabe77d9961c5487347a2170c78ac77770d8af)

Ah, I traced `LLVM_PREFER_STATIC_ZSTD` to https://github.com/llvm/llvm-project/commit/c0b4f248df79f184adba856f13a950a53c881f3f and assumed it didn't make it into LLVM 15 (it was introduced after `15.x` branched) but it turns out that [this](https://github.com/llvm/llvm-project/commit/55d4d86e388514f8ff5872e08f0d1e5d6865acdb) and the commit changing it to [`LLVM_USE_STATIC_ZSTD`](https://github.com/llvm/llvm-project/commit/4bd3f3759259548e159aeba5c76efb9a0864e6fa) were backported to LLVM 15.

So we can just mirror the change you made to `llvmPackages_16` exactly.

File: pkgs/development/compilers/llvm/15/llvm/default.nix [hunk 3]
```diff
@@ -329,6 +331,8 @@ in stdenv.mkDerivation (rec {
     # file and doesn't link zlib as well.
     # https://github.com/ClangBuiltLinux/tc-build/issues/150#issuecomment-845418812
     "-DLLVM_ENABLE_LIBXML2=OFF"
+  ] ++ optionals enableZstd [
+    "-DLLVM_PREFER_STATIC_ZSTD=${toString stdenv.hostPlatform.isStatic}"
```
Comment (rrbutani) [inline pkgs/development/compilers/llvm/15/llvm/default.nix:335]: As noted, the `LLVM_USE_STATIC_ZSTD` change was backported to LLVM 15: https://github.com/llvm/llvm-project/commit/4bd3f3759259548e159aeba5c76efb9a0864e6fa

```suggestion
    # We can use this option to make llvm to use static version of zstd.
    #
    # This option was backported to LLVM 15: https://github.com/llvm/llvm-project/commit/4bd3f3759259548e159aeba5c76efb9a0864e6fa
    "-DLLVM_USE_STATIC_ZSTD=${toString stdenv.hostPlatform.isStatic}"
```

**Review by rrbutani — CHANGES_REQUESTED** [review]
Looks good! Just one last set of nits.

File: nixos/modules/services/printing/cupsd_tmpfiles.conf
Comment (SaltyKitkat) [reply nixos/modules/services/printing/cupsd_tmpfiles.conf:1]: exactly！

**Comment by nh2** [comment]
> given the direction #237563 is headed

FYI, moved to #243161

**Comment by SaltyKitkat** [comment]
ping @rrbutani
Can this be merged now?
Or need more improvements?

**Comment by rrbutani** [comment]
Hey @SaltyKitKat, I'm sorry this was dropped.

Now that #258614 is merged can you please update this PR to apply the same changes to `llvmPackages_17` and `llvmPackages_git`?

I'll try to rebase and test that these changes are still good when applied to up-to-date `staging` today evening; if everything builds we can merge this.

**Comment by drupol** [comment]
I'm locking and closing this pull request, because through the rebase you inadvertently requested many maintainers for review, which subscribed them to update notifications, resulting in unnecessary spam inside everyone's inbox. While we can remove their review requests, we sadly cannot unsubscribe anyone.

Please create a new pull request and for the next time remember to set your PR to draft status before rebasing. In draft status, you can preview the list of maintainers that are about to be requested for review, which allows you to sidestep this issue.

**Pull request closed without merging**